### DISCUSSION TITLE
HQL syntax sugar for array operations

### DISCUSSION THREAD

**Discussion by beikov** [hibernate/hibernate-orm#7482] [Design Proposals]
Since we implemented array functions now [as discussed](https://github.com/hibernate/hibernate-orm/discussions/5562) via https://github.com/hibernate/hibernate-orm/pull/7392, https://github.com/hibernate/hibernate-orm/pull/7462 and https://github.com/hibernate/hibernate-orm/pull/7480, we were thinking about making it easier for people to work with arrays.

One such simplification was by implementing syntax sugar for concatenating arrays with the `||` (double pipe) operator.

@gavinking [brought up](https://github.com/hibernate/hibernate-orm/discussions/5562#discussioncomment-7402406) some other possible simplifications. I'll list them here and we can discuss these simplifications:

* Array construction short-hand by using brackets `[elem1, elem2, ...]`
* Support arrays on the RHS of the `in` predicate
* Access array element by index via bracket index syntax i.e. `array[1]`
* Overload/add functions without the `array_` prefix i.e. `length()` for arrays

Some proposals which I don't think will fly are the following:

* Overload `concat`. I don't think it's worth it, since we have support for the `||` operator already. I think this would only complicate the rendering/construction in the `concat` function, but we can certainly try this.
* Setting elements of an array via bracket index syntax. How should that work or look i.e. how to differentiate it from a comparison?

Here are additional some ideas I have for simplification:

* Introduce `overlaps` and `contains` predicates i.e. `where e.array contains 'abc'`. These can be used for [range type support](https://github.com/hibernate/hibernate-orm/discussions/7070) and Hibernate Spatial as well.
* Add slicing syntax like PostgreSQL i.e. `array_slice(array, 1, 10)` => `array[1:10]`

Here is a list of the functions we have or plan to have:

| Function | Syntax sugar |
|--------|--------|
| `array()` | Short-hand bracket `[elem1, elem2]` |
| `array_agg()` | _none_ |
| `array_position()` | Overload `position()` |
| `array_positions()` | _none_ |
| `array_length()` | Overload `length()` |
| `array_concat()` | Double-pipe operator |
| `array_prepend()` | Double-pipe operator |
| `array_append()` | Double-pipe operator |
| `array_contains()` | Custom `contains` predicate |
| `array_contains_nullable()` | _none_ |
| `array_overlaps()` | Custom `overlaps` predicate |
| `array_overlaps_nullable()` | _none_ |
| `array_get()` | Bracket index syntax `array[1]` |
| `array_set()` | _none_ |
| `array_remove()` | _none_ |
| `array_remove_index()` | _none_ |
| `array_slice()` | PostgreSQL slicing syntax `array[1:10]` |
| `array_replace()` | _none_ |
| `array_trim()` | _none_ |
| `array_fill()` | _none_ |
| `array_to_string()` | _none_ |

Thoughts or concerns with the presented syntax sugar? Ideas about further syntax sugar? Let me know in the comments!

**Comment by gavinking** [comment]
> How should that work or look i.e. how to differentiate it from a comparison?

To be clear I have not thought it through, but I was thinking of something like this:

    update Entity set itsArray[0] = :theElement

> `array_to_string()`

I think it would be much nicer if this was handled by the existing `str()` function. (Or even by `cast(array as String)`.)

**Reply by beikov** [reply]
>To be clear I have not thought it through, but I was thinking of something like this:

I certainly would like to support this syntax in HQL/Criteria, but know that the `array_set` function is also "useful" if you want to produce a new array with an element set a specified index, so not _just_ for actually updating an array column.

>I think it would be much nicer if this was handled by the existing str() function. (Or even by cast(array as String).)

Steve actually said something similar, but the function also takes a `separator` argument 😄

**Reply by gavinking** [reply]
> Steve actually said something similar, but the function also takes a separator argument 😄

Well, sure, but we have already taken the plunge on allowing functions to be overloaded, so I guess that's fine.

**Comment by beikov** [comment]
I created https://hibernate.atlassian.net/browse/HHH-17375 for this work. Further comments/questions/suggestions can happen on the Jira issue.
